core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -69,10 +108,23 @@ class KafkaRequestHandler(id: Int,
           completeShutdown()
           return
 
+        case request: RequestChannel.CallbackRequest =>
+          try {
+            request.originalRequest.callbackRequestDequeTimeNanos = Some(time.nanoseconds())
+            request.fun() 
+            request.originalRequest.callbackRequestCompleteTimeNanos = Some(time.nanoseconds())
```

Comment:
I think the metric could be updated when the response is sent in request.fun(), so it may actually see some value in request.originalRequest.callbackRequestDequeTimeNanos and no value in callbackRequestCompleteTimeNanos, in which case the processing time could be negative.

Also, if multiple callbacks are used during the request processing, we'd lose callback processing time for all but last callback (for this we could add a comment, to add this logic if we have multiple callbacks to handle).

Reply:
Are you saying network client sends the response before we reach line 115, so we should put the complete message as part of the response? I'm not sure the best way to put that in the callback. I'll have to think about it.

Reply:
Ah I think I figured it out.